github/messages.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2016 The go-github AUTHORS. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// This file provides functions for validating payloads from GitHub Webhooks.
+// GitHub docs: https://developer.github.com/webhooks/securing/#validating-payloads-from-github
+
+package github
+
+import (
+	"crypto/hmac"
+	"crypto/sha1"
+	"crypto/sha256"
+	"crypto/sha512"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"hash"
+	"io/ioutil"
+	"net/http"
+	"strings"
+)
+
+const (
+	// sha1Prefix is the prefix used by GitHub before the HMAC hexdigest.
+	sha1Prefix = "sha1"
+	// sha256Prefix and sha512Prefix are provided for future compatibility.
+	sha256Prefix = "sha256"
+	sha512Prefix = "sha512"
+	// signatureHeader is the GitHub header key used to pass the HMAC hexdigest.
+	signatureHeader = "X-Hub-Signature"
+)
+
+// genMAC generates the HMAC signature for a message provided the secret key and hashFunc.
+// If hashFunc is nil, sha1.New is used.
+func genMAC(message, key []byte, hashFunc func() hash.Hash) []byte {
+	if hashFunc == nil {
+		hashFunc = sha1.New
+	}
```

Comment:
Style, since `checkMAC` is the only caller of `genMAC`, I think `genMAC` can be simplified by factoring the "If hashFunc is nil, sha1.New is used." behavior out to `checkMAC`. I think this is simpler, isn't it?

```Go
// genMAC generates the HMAC signature for a message provided the secret key and hashFunc.
func genMAC(message, key []byte, hashFunc func() hash.Hash) []byte {
	mac := hmac.New(hashFunc, key)
	mac.Write(message)
	return mac.Sum(nil)
}

// checkMAC reports whether messageMAC is a valid HMAC tag for message.
// If hashFunc is nil, sha1.New is used.
func checkMAC(message, messageMAC, key []byte, hashFunc func() hash.Hash) bool {
	if hashFunc == nil {
		hashFunc = sha1.New
	}
	expectedMAC := genMAC(message, key, hashFunc)
	return hmac.Equal(messageMAC, expectedMAC)
}
```